src/main/java/org/radarcns/management/web/rest/TokenKeyEndpoint.java

```diff
@@ -0,0 +1,53 @@
+package org.radarcns.management.web.rest;
+
+import java.security.Principal;
+import java.util.List;
+import javax.annotation.PostConstruct;
+
+import com.codahale.metrics.annotation.Timed;
+import org.radarcns.auth.exception.NotAuthorizedException;
+import org.radarcns.management.config.ManagementPortalProperties;
+import org.radarcns.management.security.jwt.ManagementPortalOauthKeyStoreHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.authentication.AnonymousAuthenticationToken;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class TokenKeyEndpoint {
+
+    private static final Logger logger = LoggerFactory.getLogger(TokenKeyEndpoint.class);
+
+    @Autowired
+    private ManagementPortalProperties managementPortalProperties;
+
+    private ManagementPortalOauthKeyStoreHandler managementPortalOauthKeyStoreHandler;
+
+    @PostConstruct
+    public void init() {
+        this.managementPortalOauthKeyStoreHandler =
+                ManagementPortalOauthKeyStoreHandler.build(managementPortalProperties);
+    }
+
+    /**
+     * Get the verification key for the token signatures. The principal has to
+     * be provided only if the key is secret
+     * (shared not public).
+     *
+     * @param principal the currently authenticated user if there is one
+     * @return the key used to verify tokens
+     */
+    @GetMapping("/oauth/token_key")
+    @Timed
+    public List<String> getKey(Principal principal) throws NotAuthorizedException {
+        logger.debug("Requesting verifier public keys...");
+        if ((principal == null || principal instanceof AnonymousAuthenticationToken)) {
+            throw new NotAuthorizedException("You need to authenticate to see a shared key");
+        }
+
+        return managementPortalOauthKeyStoreHandler.loadVerifyingPublicKeys();
```

Comment:
If we're going to expose multiple web keys, why not use the JWKS spec: https://auth0.com/docs/jwks
Detailed JSON: https://auth0.com/docs/tokens/reference/jwt/jwks-properties

Reply:
Example from the spec:
```JSON
{"keys":
       [
         {"kty":"EC",
          "crv":"P-256",
          "x":"MKBCTNIcKUSDii11ySs3526iDZ8AiTo7Tu6KPAqv7D4",
          "y":"4Etl6SRW2YiLUrN5vfvVHuhp7x8PxltmWWlbbM4IFyM",
          "use":"enc",
          "kid":"1"},

         {"kty":"RSA",
          "n": "0vx7agoebGcQSuuPiLJXZptN9nndrQmbXEps2aiAFbWhM78LhWx
     4cbbfAAtVT86zwu1RK7aPFFxuhDR1L6tSoc_BJECPebWKRXjBZCiFV4n3oknjhMs
     tn64tZ_2W-5JsGY4Hc5n9yBXArwl93lqt7_RN5w6Cf0h4QyQ5v-65YGjQR0_FDW2
     QvzqY368QQMicAtaSqzs8KJZgnYb9c7d0zgdAZHzu6qMQvRL5hajrn1n91CbOpbI
     SD08qNLyrdkt-bFTWhAI4vMQFh6WeZu0fM4lFd2NcRwr3XPksINHaQ-G_xBniIqb
     w0Ls1jF44-csFCur-kEgU8awapJzKnqDKgw",
          "e":"AQAB",
          "alg":"RS256",
          "kid":"2011-04-29"}
       ]
     }
```

Reply:
So I had a look into JWKS spec and possibilities to implement it on MP. 
From what I read every public-key needs a unique-id as an identifier and it is added as 'kid' in the header. I do not think it is the case right now. But probably there is a way to do that easily. However, the bigger confusion is current implementation just needs the public-key to create a the algorithm the the verifier. So we have to change the implementation to use JWKProvider. 
Plus oauth0, has jwk library for [rsa](https://github.com/auth0/jwks-rsa-java) only. I think we need both rsa and EC.
Other option is to use [Nimbus JOSE+JWT library ](https://connect2id.com/products/nimbus-jose-jwt/examples/jwk-retrieval), assuming we will replace the use of oauth0 with Nimbus. What do you think? @blootsvoets 
Am I missing something?

Reply:
One implementation of the `kid` is to use the SHA-1 sum of the key. This way no additional bookkeeping is needed. If auth0 doesn't support it, eventually probably Nimbus would be a better choice. For now, we can stick to auth0, but perhaps do match the general layout:
```json
{"keys":
       [
         {
          "alg": <algorithm>,
          "value": <Auth0 encoded public key>
         }
       ]
}
```

Reply:
Make sense. I agree. For now I will change the API to fit what is proposed here. Later, we can investigate on migrating to Nimbus. I have one last question. Just to confirm.
Should the `oauth/token_key` an openly available endpoint i.e. no authentication? I think it should not be. Currently we expect some valid Principal to be available. 
Assuming only registered clients will request for JWKs, can I use basic auth with the client-credentials to request JWK? @blootsvoets

Reply:
In OpenID connect, public keys are considered public, so no authentication should be necessary. This is generally true, public keys can always be exposed.